galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunArtifactsDownloadRoute.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright contributors to the Galasa project
+ */
+package dev.galasa.framework.api.ras.internal;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.ByteChannel;
+import java.nio.file.FileSystem;
+import java.nio.file.OpenOption;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.nio.file.attribute.FileAttribute;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletResponse;
+
+import com.google.gson.Gson;
+
+import dev.galasa.framework.IFileSystem;
+import dev.galasa.framework.spi.FrameworkException;
+import dev.galasa.framework.spi.IFramework;
+import dev.galasa.framework.spi.IRunResult;
+import dev.galasa.framework.spi.ResultArchiveStoreException;
+import dev.galasa.framework.spi.utils.GalasaGsonBuilder;
+
+import static dev.galasa.framework.api.ras.internal.ServletErrorMessage.*;
+
+public class RunArtifactsDownloadRoute extends RunsRoute {
+    
+    static final Gson gson = GalasaGsonBuilder.build();
+
+    public RunArtifactsDownloadRoute(IFileSystem fileSystem, IFramework framework) {
+        super("\\/runs\\/([A-z0-9.\\-=]+)\\/files\\/([A-z0-9.\\-=\\/]+)", fileSystem, framework);
+    }
+
+    @Override
+    public HttpServletResponse handleRequest(String pathInfo, QueryParameters queryParams, HttpServletResponse response) throws ServletException, IOException, FrameworkException {
+        Matcher matcher = Pattern.compile(this.getPath()).matcher(pathInfo);
+        matcher.matches();
+        String runId = matcher.group(1);
+        String artifactPath = matcher.group(2);
+        return downloadArtifact(runId, artifactPath, response);
+    }
+
+    private HttpServletResponse downloadArtifact(String runId, String artifactPath, HttpServletResponse res) throws InternalServletException, IOException {
+        IRunResult run = null;
+        String runName = "";
+        String artifactsPrefix = "artifacts/";
+        
+        // Get run details in order to find artifacts
+        try {
+            run = getRunByRunId(runId);
+            runName = run.getTestStructure().getRunName();
+        } catch (ResultArchiveStoreException e) {
+            ServletError error = new ServletError(GAL5002_INVALID_RUN_ID,runId);
+            throw new InternalServletException(error, HttpServletResponse.SC_NOT_FOUND);
+        }
+        
+        Map<String, IRunRootArtifact> rootArtifacts = new HashMap<>();
```

Comment:
build the map in the constructor ? Then it only needs building once ?

Reply:
Addressed in https://github.com/galasa-dev/framework/pull/403/commits/a5ff45f32baa087c11b79f09319d9ced4d86a09e